Improved app user support

cburnette opened the pull request:
Box iOS team, please review.  Thanks!

cburnette committed:
debug message test

cburnette committed:
debug message back to normal

cburnette committed:
added new method to set access token directly

cburnette committed:
added public reference to setAccessToken

cburnette committed:
adding support for creating a BoxContentClient wrapper that uses server auth to retrieve tokens and can be instantiated with only a token and a user id; supports app users, token downscoping, service accounts, etc.

cburnette committed:
removed unneeded method

cburnette committed:
added new header to target

cburnette committed:
had to make header public

cburnette committed:
renamed param

cburnette committed:
added two assert checks

cburnette committed:
changed parameter names and added a userInfo dictionary for relevant information about the user (scopes, etc.)

cburnette committed:
removed api call to retrieve user after fetching a new token.  This is not required for the server auth model and it can fail due to a downscoped token that doesn't have rights to read user info.

cburnette committed:
reordered parameter list

cburnette committed:
added a typedef for the fetchToken signature

cburnette committed:
added some property definitions to help with Swift interfaces.  No longer requiring userId to be nonnull.

cburnette committed:
slight code tweak

cburnette committed:
added new doc file for Box Platform; removed old AppUsers doc (which was only partially correct and is now out of date)

cburnette committed:
doc tweak

cburnette committed:
updated to reflect new app patterns

cburnette committed:
doc tweak

boxcla commented:
Verified that @cburnette has signed the CLA. Thanks for the pull request!

jlawton commented on BoxContentSDK/BoxContentSDK/Clients/BoxContentClientServerAuthWrapper.h:
```diff
@@ -0,0 +1,18 @@
+#import "BOXContentClient.h"
+#import "BOXAPIAccessTokenDelegate.h"
+
+typedef void (^ServerAuthFetchTokenBlock)(NSString *userId, NSDictionary *userInfo, void (^)(NSString *, NSDate *, NSError *));
```
Suggestion: This is a little hard to read, especially without names on the callback parameters. I might consider this or similar as an option:

```
typedef void (^ServerAuthFetchTokenBlock)(NSString *userId, NSDictionary *userInfo, TokenFetchCompletion completion);

// Either provide a token and its expiration, or an error.
typedef void (^TokenFetchCompletion)(NSString * __nullable accessToken, NSDate * __nullable tokenExpiration, NSError * __nullable error);
```

jlawton commented on BoxContentSDK/BoxContentSDK/Clients/BoxContentClientServerAuthWrapper.h:
```diff
@@ -0,0 +1,18 @@
+
+@property (nonatomic, nonnull, readonly, strong) BOXContentClient *boxClient;
+@property (nonatomic, nullable, readonly, strong) NSString *userId;
+@property (nonatomic, nullable, readwrite, strong) NSDictionary *userInfo;
```
NSString and NSDictionary should be copied, lest they be modified under us.

jlawton commented on BoxContentSDK/BoxContentSDK/Clients/BoxContentClientServerAuthWrapper.h:
```diff
@@ -0,0 +1,18 @@
+#import "BOXContentClient.h"
+#import "BOXAPIAccessTokenDelegate.h"
+
```
Wrap body of file in `NS_ASSUME_NONNULL_BEGIN` and `NS_ASSUME_NONNULL_END` and we only need to call out `nullable` elsewhere.

jlawton commented on BoxContentSDK/BoxContentSDK/Clients/BoxContentClientServerAuthWrapper.m:
```diff
@@ -0,0 +1,36 @@
+
+@interface BoxContentClientServerAuthWrapper ()
+
+@property (nonnull, nonatomic, readonly, strong) ServerAuthFetchTokenBlock fetchTokenBlock;
```
Blocks should be copied or we could be given a stack allocated block and crash.

jlawton commented on BoxContentSDK/BoxContentSDK/Clients/BoxContentClientServerAuthWrapper.m:
```diff
@@ -0,0 +1,36 @@
+#import "BoxContentClientServerAuthWrapper.h"
+#import "BoxAbstractSession.h"
+
```
Also assume nonnull in here.

jlawton commented on BoxContentSDK/BoxContentSDK/Clients/BoxContentClientServerAuthWrapper.h:
```diff
@@ -0,0 +1,18 @@
+@property (nonatomic, nullable, readwrite, strong) NSDictionary *userInfo;
+
+
+- (instancetype)initWithToken:(nullable NSString*)token
```
Is it okay to take a token without the ability to take an expiration date?

jlawton commented on doc/BoxPlatform.md:
```diff
@@ -0,0 +1,74 @@
+
+In order to support this application architecture, there is a wrapper class included in the SDK to handle the mechanics of retrieving tokens for a user from the remote server. It includes the ability to specify an initial token that you may have already retrieved, and allows you to specify a code block that will be called by the SDK when it detects an expired token.  This code block is where you  ...
+
+<b>PLEASE NOTE:</b> It is up to you as the app developer to ensure that the process of retrieving tokens is secured.  Typically this is done by passing a JSON Web Token in the Authorization header of the request to your remote server that is signed by your authentication system.  That way you can verify that the incoming request is indeed originating from an authorized app and device.
```
I know that this is more of a side note, but I'm curious how you securely get a JWT onto the user's device, which is what I think I'm reading here.

cburnette replied:
As an example, in the app I put together for Boxworks that uses AWS Mobile Hub for authentication, there is a secure connection established between the app and AWS using a secret (and their SDK), and when a user provides username/password and its validated, AWS generates a signed JWT and returns it to the app.  That JWT is send in the Authentication header for all the other calls to other AWS services, like Lambda, where the Box token generation happens.

jlawton commented on BoxContentSDK/BoxContentSDK/Clients/BoxContentClientServerAuthWrapper.m:
```diff
@@ -0,0 +1,36 @@
+
+- (instancetype)initWithToken:(nullable NSString *)token
+                       userId:(nullable NSString *)userId
+                     userInfo:(nullable NSDictionary *)userInfo
```
Why have a `userInfo` when the block could just close around what it needs? I suppose because it's a readwrite property? If that's the rationale, it should definitely be atomic, too.

jlawton commented on doc/BoxPlatform.md:
````diff
@@ -0,0 +1,74 @@
+
+Below is an example of configuring the <b>BoxContentClientServerAuthWrapper</b> instance to work with a specific App User Id, including example code of how to retrieve a new token when the SDK detects a token expiration. Note that you are not require to specify an initial token as the SDK will simply invoke your fetchTokenBlock. However, in some cases you may already have a token (for example, a  ...
+
+```objectivec
````
We get a lot of complaints that the SDK is Objective-C, and users want Swift. There are good reasons not to write the SDK in Swift yet, but maybe we do want our sample code to be Swift? I can see arguments that we wouldn't want that, but thought it should be considered.

cburnette replied:
good point, I have Swift example code I can include

jlawton commented on BoxContentSDK/BoxContentSDK/Clients/BoxContentClientServerAuthWrapper.m:
```diff
@@ -0,0 +1,36 @@
+
+- (void)fetchAccessTokenWithCompletion:(void (^)(NSString *, NSDate *, NSError *))completion
+{
+    _fetchTokenBlock(_userId, _userInfo, completion);
```
I don't see a lot of code directly accessing ivars these days. Might want to use `self.fetchTokenBlock`

jlawton commented on BoxContentSDK/BoxContentSDK/OAuth2/BOXAppUserSession.m:
```diff
@@ -24,25 +24,9 @@ - (void)performAuthorizationCodeGrantWithReceivedURL:(NSURL *)URL withCompletion
-        BOXUserRequest *userRequest = [[BOXUserRequest alloc] init];
-        [weakSelf prepareRequest:userRequest];
-        
-        [userRequest performRequestWithCompletion:^(BOXUser *user, NSError *error) {
```
Removing this is not a backwards-compatible change for the non-app-user case, I believe. How do we at least get our automatic keychain storage to work for the normal Box user case?

cburnette replied:
@jlawton  so actually it is (I believe) as that session is only used for app user sessions.  The other sessions are un-modified.  Anyway, I just checked in a big change where I realized I can eliminate the wrapper and just move this functionality into BOXContentClient with a new initializer class method.  Have a look and let me know what you think.  thanks!

cburnette replied:
@jlawton  I don't believe I've changed anything with the existing user models

jlawton replied:
Ah. Great. I'll take another look.

jlawton reviewed (commented):
Sorry that these comments are at a couple of levels of abstraction. At a high level, I think this direction is getting there, but I worry that this change doesn't maintain backwards compatibility. For instance, I know that it breaks assumptions in the official Box iOS app.

cburnette committed:
experimenting moving the code into BOXContentClient instead of having an additional wrapper

cburnette committed:
checking for null fetchTokenBlock

cburnette committed:
rearranged assignments

cburnette committed:
realized I can eliminate the wrapper and accomplish the same results in BOXContentClient

cburnette committed:
doc tweak

cburnette committed:
added comments for new properties and method

jlawton commented on BoxContentSDK/BoxContentSDK/Clients/BOXContentClient.h:
```diff
@@ -74,6 +77,18 @@ extern NSString *const BOXContentClientBackgroundTempFolder;
+ * When using a server-based auth client (App Users or downscoped tokens) this property store the box user id of the associated user.
+ * Passed into your fetchTokenBlock delegate method when a new token is required.
+ */
+@property (nonatomic, nullable, readwrite, strong) NSString *userId;
```
I see a lot of potential for using this accidentally, or setting it at the wrong time. It should probably only be in the implementation file, and if the developer wants to switch users, they should be forced to create a fresh client. (Should also be a `copy` property.)

cburnette replied:
good point, i'll make that change.

cburnette replied:
and yes, my intention is that if you want a client for a different app user for some reason you would create a new one.

jlawton commented on BoxContentSDK/BoxContentSDK/Clients/BOXContentClient.h:
```diff
@@ -74,6 +77,18 @@ extern NSString *const BOXContentClientBackgroundTempFolder;
+ * This property is for storing any relevant information needed when your fetchTokenBlock delegate method is invoked
+ * to retrieve a new token.  Passed into your fetchTokenBlock delegate method when a new token is required.
+ */
+@property (nonatomic, nullable, readwrite, strong) NSDictionary *userInfo;
```
Should also be an `atomic` and `copy` property. I wonder if the name should be more specific.

cburnette replied:
so on the atomic vs. nonatomic thing, i looked through the existing code and it seems like most things should be atomic but are nonatomic so I went with that.  I can change to atomic.  as for userInfo I went with that name as that's the idiomatic name for a dictionary of random information to be delivered in a callback, correct?

jlawton replied:
It's no doubt true that we're wrong elsewhere. This is presumably expected to be written by the user, and also used in a callback on an internal queue, so it seems we should get the semantics right for this. Most of our readwrite properties are really only expected to be written in some sort of setup phase. If that were the case here, the user could just capture the value in their block.

I'm going back and forth on the name. `userInfo` is idiomatic, as you say. The trouble I'm having is that calling this particular block I wouldn't really consider the main responsibility of the class, so the generic name just feels odd. It's probably just me, though.

cburnette replied:
As far as the name, I'm open to suggestions.  In terms of capturing this info in the block, keep in mind that typically the way this is set up there is one block that contains the logic to call the backend given a few parameters, like a user id and any extra info like a scope list.  So you could instantiate multiple BOXContentClients that all have the same fetchTokenBlock but have different user ids and extra user info.

cburnette replied:
what about fetchTokenBlockInfo?

jlawton commented on BoxContentSDK/BoxContentSDK/Clients/BOXContentClient.m:
```diff
@@ -46,6 +46,8 @@ @interface BOXContentClient ()
 
 @property (nonatomic, readwrite, strong) BOXSharedLinkHeadersHelper *sharedLinksHeaderHelper;
 
+@property (nonnull, nonatomic, readwrite, strong) ServerAuthFetchTokenBlock fetchTokenBlock;
```
This should be a `copy` property, or we're inviting crashes.

cburnette replied:
ok cool, didn't realize you could copy a block.  Thanks!

jlawton reviewed (commented):
I like this version better than the previous one.

cburnette committed:
updates to property attributes; moved userId to internal property

cburnette committed:
throw exception if fetchAccessTokenWithCompletion is invoked and no fetchTokenBlock has been specified.  Highly unlikely to occur but just in case.

cburnette committed:
just moving the star

cburnette committed:
removed unneeded parameter names in typedef;
renamed userInfo to fetchTokenBlockInfo to be a bit more specific

cburnette commented:
Please don't anyone merge this yet as I'm working through a nuance when using this with the Preview SDK.  Thanks!

cburnette committed:
added notification of session refresh back (got a new token from the server;
changed method signature for creating a server auth client

cburnette committed:
requiring unique id

cburnette committed:
public method declaration fix

cburnette committed:
updated param comment

cburnette committed:
updated doc file

cburnette commented:
reopened with adjusted design

cburnette committed:
added uniqueIdentifier property.

cburnette committed:
comments around uniqueIdentifier

cburnette committed:
no need for custom error

cburnette committed:
Introduce protocol UniqueSDKUser;
Modified BOXUserMini to conform to UniqueSDKUser protocol;
Converted former BOXUserMini properties to return id<UniqueSDKUser> in BOXContentClient and BOXAbstractSession;
Refactored all references to user.modelId to be user.uniqueId;

cburnette committed:
switched fetchTokenBlock to use UniqueSDKUser protocol property uniqueId

cburnette committed:
updated to use UniqueSDKUser protocol and property uniqueId

cburnette committed:
updated remaining BOXUserMini references in the client and abstract session classes

cburnette committed:
removed reference to the private header; no longer needed

cburnette committed:
removed space

cburnette committed:
adding ServerAuthUser that implements protocol UniqueSDKUser; used when creating server auth client

cburnette committed:
forgot the header file

cburnette committed:
converted from BOXUserMini to UniqueSDKUser

cburnette committed:
reordered operations

cburnette committed:
restoring unique initializer for BOXUserMini.  Companion method to initializer for ServerAuthUser

cburnette committed:
changed name of methods back to what they are in Master.  No need to change them more than necessary

jlawton commented on BoxContentSDK/BoxContentSDK/Clients/BOXContentClient.h:
```diff
@@ -74,6 +79,12 @@ extern NSString *const BOXContentClientBackgroundTempFolder;
+ * This property is for storing any relevant information needed when your fetchTokenBlock delegate method is invoked
+ * to retrieve a new token.  Passed into your fetchTokenBlock delegate method when a new token is required.
+ */
+@property (nonatomic, nullable, readwrite, strong) NSDictionary *fetchTokenBlockInfo;
```
This should be `copy`, rather than `strong`

jlawton commented on BoxContentSDK/BoxContentSDK/Clients/BOXContentClient.m:
```diff
@@ -126,22 +133,32 @@ + (BOXContentClient *)clientForUser:(BOXUserMini *)user
 }
 
-+ (BOXContentClient *)clientForNewSession
++ (BOXContentClient *)clientForServerAuthUser:(nonnull ServerAuthUser*)serverAuthUser
```
Nit: Pointer spacing should be `Type *`

jlawton commented on BoxContentSDK/BoxContentSDK/Clients/BOXContentClient.m:
```diff
@@ -409,6 +434,7 @@ - (NSString*)tempCacheDir
 }
 
 - (void)setAccessTokenDelegate:(id<BOXAPIAccessTokenDelegate>)accessTokenDelegate
+                serverAuthUser:(ServerAuthUser*)serverAuthUser
```
Nit: Pointer spacing.

jlawton commented on BoxContentSDK/BoxContentSDK/Clients/BOXContentClient.h:
```diff
@@ -118,6 +129,21 @@ extern NSString *const BOXContentClientBackgroundTempFolder;
+ *
+ * @return BOXContentClient that is configured for server-based auth
+ */
++ (BOXContentClient *)clientForServerAuthUser:(nonnull ServerAuthUser*)serverAuthUser
```
Nit: Spacing.

jlawton commented on BoxContentSDK/BoxContentSDK/Models/BOXUser.h:
```diff
@@ -7,12 +7,43 @@
+
+@property (nonnull,  nonatomic, readonly, strong) NSString *uniqueId;
+@property (nullable, nonatomic, readonly, strong) NSString *name;
+@property (nullable, nonatomic, readonly, strong) NSString *login;
```
Maybe add doc comments to the protocol, as well as, or rather than, the ServerAuthUser.

jlawton replied:
These properties (and in ServerAuthUser) should be `copy` rather than `strong`.

jlawton commented on BoxContentSDK/BoxContentSDK/Models/BOXUser.m:
```diff
@@ -44,6 +68,10 @@ - (instancetype)initWithUserID:(NSString *)userID name:(NSString *)name login:(N
     return self;
 }
 
+- (NSString *)uniqueId{
```
Nit: Space before `{`

jlawton commented on BoxContentSDK/BoxContentSDK/OAuth2/BOXAbstractSession.m:
```diff
@@ -258,8 +258,8 @@ + (NSArray *)usersInKeychain
     NSArray *sortedUsers = [users sortedArrayUsingComparator:^NSComparisonResult(id a, id b) {
-        BOXUserMini *userA = (BOXUserMini *) a;
-        BOXUserMini *userB = (BOXUserMini *) b;
+        id<UniqueSDKUser> userA = (id<UniqueSDKUser>) a;
```
You can declare `a` and `b` as `id<UniqueSDKUser>` in the block parameters, and you don't need the casts.

jlawton replied:
I wonder if this sort is even necessary, but we don't have to address that here.

cburnette committed:
updates based on comments from jlawton

cburnette commented:
@jlawton Hey James, just checked in changes to address your comments. Thanks!
